apiserver/oauth-proxy-proxy.md

```diff
@@ -0,0 +1,141 @@
+The `oauth-proxy` is used to front many webservers to provide authentication and authorization protection without directly
+building it into their binaries. This is common, easy, and used by people without a deep understand of how OpenShift is
+wired together.  They are not likely to know how to reactively wire proxy configuration and ca-bundles into their 
+side-cars.  Because it is re-used by directly embedding the `oauth-proxy` into a `PodSpec`, any changes to the way it is
+invoked will cause adoption difficulties.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+In order to function, the `oauth-proxy`
+ 1. Gets the oauth-server location using the kube-apiserver's internal IP: `https://172.30.0.1/.well-known/oauth-authorization-server`
+ 2. Uses the result to find the token endpoint.
+ 3. Hits the *external* token endpoint `https://oauth-openshift.apps.<cluster-dns-prefix>.openshift.com/oauth/token`.
+
+This flow requires hitting a potentially external router, which requires honoring proxy settings.
+
+## Motivation
+
+See summary in first section.
+
+### Goals
+
+1. Zero-config proxy support that dynamically updates as proxy settings change.
+2. Support for all proxy modes, including MITM.
+
+### Non-Goals
+
+1. Making generic code to rewrite proxy transport configuration.
+
+## Proposal
+
+The `oauth-proxy` can be written to use targeted list/watch against the proxy.config.openshift.io and 
+`oc get -n openshift-config-managed configmaps/proxy-ca` to build a proxy transport.  Using standard informers with single item
+watches allows RBAC to only expose GET for individual items.  It also means that an efficient swap mechanism can be built
+to ensure performant behavior.
+
+This is an intricate thing to write, so we avoided it for every other proxy component, but this one is a component that is
+re-used by numerous external teams by directly embedding it into a pod spec.  This changes the tradeoff, so it becomes worthwhile
+to write this code.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. If HTTP env vars are set, the dynamic proxy-support code must be disabled.  The user is in charge of what the `oauth-proxy`
+ does in these cases.
+2. The list/watch driving the informers must be focused to a single item, so the RBAC special case for single item watch
+ takes effect.
+3. To be performant, the proxy and TLS configuration should be cached in something like a `sync.Atomic`.
+
+
+### Risks and Mitigations
+
+This code is intricate to write, but luckily our monitoring stack already relies on this.  If their tests pass when they
+update to use the feature, the feature works.
```

Comment:
Do we have automated tests that cover this?  I assume not, i think the MITM proxy scenario is one that's setup/run by hand by QE?

but yes, we at least have the BZ+scenario that needs to pass.

Reply:
(the e2e-aws-proxy job we have does not use a MITMing proxy as far as i know, but i could be wrong... @ewolinetz knows best)

Reply:
correct, currently we're not doing a MITM proxy for CI

Reply:
> correct, currently we're not doing a MITM proxy for CI

@ewolinetz https://github.com/openshift/release/pull/5308 will add this support, correct?

Reply:
no, that PR still only maintains a pass through proxy. It seeks to 'black hole' the private subnet though so all traffic must go through the proxy in order to reach anything outside the cluster (and anything not in the `noProxy` list)